avalon/io.py

```diff
@@ -132,9 +132,6 @@ def _from_environment():
             # Name of current app
             ("AVALON_APP", None),
 
-            # Path to working directory
-            ("AVALON_WORKDIR", None),
```

Comment:
What's the motivation here?

Reply:
The reason is that this is computed based on the other keys, the workdir is based on the current context and isn't a variable on its own. The problem with storing it as one is that it needs to be ensured to remain in-sync with context changes. Whenever the context changes you'll need to update `AVALON_WORKDIR`.

I'm having similar issues with `AVALON_SILO` - it's dependent on the current `asset`.
In my mind they are both optimizations to reduce possible queries.

We could create an `update` method that would refresh all these "computed" values.

Reply:
> We could create an update method that would refresh all these "computed" values.

Yeah. The context changing is under our control, so we can just update relevant values when that happens. For example, when one changes task or asset.

Reply:
Maybe I need to think about this some more. I'm expecting there will be quite some tools that could end up triggering context changes and I'd like to try and avoid ensuring everywhere that every session attribute remains in sync, because say this would not be "automatically refreshed somewhere" I'd have to at least do something like:

```python
# change my current session
api.Session['AVALON_ASSET'] = "villain"

# refresh silo
asset = io.find_one({"name": "villain", "type": "asset"})
api.Session['AVALON_SILO'] = asset['silo']

# refresh workdir
project = io.find_one({"type": "project"})
template = project['config']['template']['work']
api.format_template(template)
```

And this logic could be in multiple locations?
Maybe something like `api.refresh_session(session)` would make more sense?

Any ideas?

Reply:
How about a `update_asset(name)`?

Reply:
In general I don't think we can expect to change the values of Session directly. Could it be read-only?

Reply:
> In general I don't think we can expect to change the values of Session directly. Could it be read-only?

It could be I'd say, yes. 

> How about a update_asset(name)?

This might make it slow if we're doing `update_asset` then `update_task` then `update_app`?
How about an `update_session` method that would allow to update multiple things at once? 

```python
#pseudocode
def update_session(session, project=None, asset=None, task=None, app=None):
    """Update session in-place to new context"""
    
    if asset:
        assert asset_exists, "Asset must exist"
        session['AVALON_ASSET'] = asset
        
    if any((project, asset)):
        # update silo

    if any((project, asset, task, app)):
        # update workdir

api.update_session(api.Session, asset="villain", task="modeling")
```

Reply:
How about `update_current_context(context)`? Session contains a lot more than asset and task, and that way it's only applicable to the current one (which at the moment is the only one there is).

Also signals should be `mixedCase`, e.g. `currentContextUpdated`

Reply:
The problem with `context` to me is, what should it contain? Is it `asset` or `AVALON_ASSET`? And where can I find which attributes I _can_ update?

> and that way it's only applicable to the current one

Yes, but this might make the api less backwards compatible in the future when we *do* support multiple sessions to be passed around. I might be thinking to far ahead though. ;)

> Session contains a lot more than asset and task,

Correct - but it also means that if we update any of them that the updates are actually propagated I believe? Say you'd update the `AVALON_DB` attribute? What needs to happen? Will it do `io.uninstall()` and start reinstalling? Or would the user manually do that? which could be fine and more explicit? What else are you expecting to update?

Reply:
> Yes, but this might make the api less backwards compatible in the future when we do support multiple sessions to be passed around. I might be thinking to far ahead though. ;)

There will always be a current Session; it's the one Maya defaults to when booting up the Loader etc. It's forwards-compatible.

> Say you'd update the AVALON_DB attribute? What needs to happen? 

You wouldn't do it by modifying Session. You would use another `update_*` function. Session is a data container-only, read-only.

> The problem with context to me is, what should it contain? Is it asset or AVALON_ASSET? And where can I find which attributes I can update?

That's a good point..

How about..

```python
context = api.Context()

# Default values come from Session
context["asset"] = "OtherAsset"

api.update_current_context(context)
```

Not saying it's "the one". Still thinking of what's the best option.

Reply:
@mottosso How about something like this (still pseudocode):

```python
def update_current_session(context):
    """Update active session in-place to new context

    Returns:
        dict: The key and values that have been changed in the session.

    """

    # Find the changed keys
    changed = {key: value for key, value in context.items() if 
               value != Session[key]}

    # Update project (or database) directly so any updates afterwards are
    # correct
    io_reinstall = False
    for key in ["AVALON_PROJECT",
                "AVALON_DB",
                "AVALON_TIMEOUT",
                "AVALON_MONGO"]:
        if key in changed:
            # These keys will require a `io.install()` call which uses the
            # current `os.environ` state. As such we'll need to update the
            # environment
            os.environ[key] = changed[key]
            io_reinstall = True

    # Install new project database
    if io_reinstall:
        io.uninstall()
        io.install()

    # Update silo when asset changes
    if "AVALON_ASSET" in changed:
        # Update silo for the asset
        asset_document = io.find_one({"name": changed["AVALON_ASSET"],
                                      "type": "asset"},
                                     projection={"silo": True})
        assert asset_document, "Asset does not exist"

        # Store our change to AVALON_SILO
        changed["AVALON_SILO"] = asset_document['silo']

    # Update work directory
    if any(key in changed for key in ["AVALON_PROJECT",
                                      "AVALON_ASSET",
                                      "AVALON_TASK",
                                      "AVALON_APP",
                                      "AVALON_USER"]):
        project = io.find_one({"type": "project"},
                              projection={"config.template.work": True})
        template = project['config']['template']['work']
        changed['AVALON_WORKDIR'] = format_template(template, Session)

    # Trigger updates on the Session
    Session.update(changed)

    return changed
```

Reply:
Wuw, that's much too complex! Also we shouldn't implicitly install/uninstall things. Those are heavy operations that do a lot of stuff and should only be done once. Think of it like installing software onto disk.

Also `context` doesn't include `AVALON_`, that's a prefix to distinguish variables in the environment. Internally, it's just `asset` etc.

I don't think we should go for `update_session`, I expect parts of it to be updated independently, especially tasks, then assets and lastly projects. I don't expect databases to change very often at all.

Maybe your expectations are different? Tell me what you've got in mind.

Reply:
Expecting to:
- At any time in an application to switch task or asset; this is one of the most requested features in our studio.
- Switch user maybe?

I would avoid to:
- Change project (whilst in application) - changing project would just require you to open up the application for that project which makes more sense to me.
- Change database or anything that would require a reinstall or changes global settings, like sentry paths, database names, etc.

As such I'm only expecting to change things like asset, task, app, user to change. (And with that to update `AVALON_SILO` and `AVALON_WORKDIR` to change too.

Reply:
Also, some of the members of Session can't change independently, now that I think of it.

Changing `project` would invalidate `asset`, along with `silo` as you mentioned. So we should make sure updates happen with all dependencies at once. Passing `context` would guarantee that. `update_session` would too, but is too heavy handed.

And unless it wasn't clear, `context` in this.. context (pun!) is this fella:

- https://github.com/getavalon/core/blob/5c0ba8d613db056a549cc4709ab7a8393627491e/avalon/maya/compat.py#L75

Also found here:

- https://github.com/getavalon/core/blob/9d8015b2798c9469e0316a2bfce78bd0d44bace0/avalon/schema/representation-2.0.json#L57

We could potentially break this out into an individual schema.

> At any time in an application to switch task or asset; this is one of the most requested features in our studio. Switch user maybe?

Yeah, that's my impression too. Let's focus on these three for now.

Reply:
Sure, we could do this:

```python
def update_current_session(asset=None,
                           task=None,
                           app=None,
                           user=None):
    """Update active session in-place to new context"""

    mapping = {
        "AVALON_ASSET": asset,
        "AVALON_TASK": task,
        "AVALON_APP": app,
        "AVALON_USER": user
    }
    changed = {key: value for key, value in mapping.items() if
               Session[key] != value}
    if not changed:
        return

    # Update silo when asset changed
    if "AVALON_ASSET" in changed:
        # Update silo for the asset
        asset_document = io.find_one({"name": asset,
                                      "type": "asset"},
                                     projection={"silo": True})
        assert asset_document, "Asset must exist"
        changed["AVALON_SILO"] = asset_document['silo']

    # Update work directory
    project = io.find_one({"type": "project"},
                          projection={"config.template.work": True})
    template = project['config']['template']['work']
    changed['AVALON_WORKDIR'] = format_template(template, Session)

    Session.update(changed)

    return changed
```

Or do the same whilst having a single argument like `context`?

This one could currently also be used like:

```update_current_session(**context)``` 
(if the context only has the allowed keys that can be changed)

Reply:
The "delta" implementation is cool, but I think we should stick with something simple for a 1.0.

Let's assume that making a change to any member of `context` updates the Session in the same way, and performs the same database operations.

`user` isn't part of `context`, let's make that separate. It's safe to assume you wouldn't want to change Asset and User at the same time.

Here's my suggestion.

```python
context = api.Context()
context["task"] = "animation"
api.update_current_context(context)

def update_current_context(context):
  workdir = compute_workdir(...)

  assert all_ok

  Session["AVALON_ASSET"] = context["asset"]
  Session["AVALON_TASK"] = context["task"]
  ...
```